pkg/services/annotations/annotationsimpl/xorm_store.go

```diff
@@ -371,64 +366,37 @@ func (r *xormRepositoryImpl) Get(ctx context.Context, query *annotations.ItemQue
 	return items, err
 }
 
-type acFilter struct {
-	where       string
-	whereParams []interface{}
-	recQueries  string
-	recParams   []interface{}
-}
-
-func (r *xormRepositoryImpl) getAccessControlFilter(user identity.Requester) (acFilter, error) {
-	var recQueries string
-	var recQueriesParams []interface{}
+func (r *xormRepositoryImpl) getAccessControlFilter(user identity.Requester, accessResources *accesscontrol.AccessResources) (string, error) {
+	var filters []string
 
-	if user == nil || user.IsNil() {
-		return acFilter{}, errors.New("missing permissions")
+	if _, ok := accessResources.ScopeTypes[annotations.Organization.String()]; ok {
+		filters = append(filters, "a.dashboard_id = 0")
 	}
 
-	scopes, has := user.GetPermissions()[ac.ActionAnnotationsRead]
-	if !has {
-		return acFilter{}, errors.New("missing permissions")
-	}
-	types, hasWildcardScope := ac.ParseScopes(ac.ScopeAnnotationsProvider.GetResourceScopeType(""), scopes)
-	if hasWildcardScope {
-		types = map[interface{}]struct{}{annotations.Dashboard.String(): {}, annotations.Organization.String(): {}}
-	}
+	if _, ok := accessResources.ScopeTypes[annotations.Dashboard.String()]; ok {
+		var dashboardIDs []int64
+		for _, id := range accessResources.Dashboards {
+			dashboardIDs = append(dashboardIDs, id)
+		}
 
-	var filters []string
-	var params []interface{}
-	for t := range types {
-		// annotation read permission with scope annotations:type:organization allows listing annotations that are not associated with a dashboard
-		if t == annotations.Organization.String() {
-			filters = append(filters, "a.dashboard_id = 0")
-		}
-		// annotation read permission with scope annotations:type:dashboard allows listing annotations from dashboards which the user can view
-		if t == annotations.Dashboard.String() {
-			recursiveQueriesAreSupported, err := r.db.RecursiveQueriesAreSupported()
-			if err != nil {
-				return acFilter{}, err
-			}
+		var inClause string
+		if len(dashboardIDs) == 0 {
+			inClause = "SELECT * FROM (SELECT 0 LIMIT 0) tt" // empty set
```

Comment:
Nit: I think we could just skip the dashboard ID filter if the user doesn't have access to reading dashboard annotations.

Reply:
This is already what happens, this code path is only reached if they have access to read dashboard annotations.

Reply:
Sorry, I meant if the `dashboardIDs` list is empty, so they don't have access to the dashboards. But I'm also ok with leaving it as it is.

Reply:
Ah no we still need a way to filter out annotations. If we just skip the filter then they'd be able to read any annotation as long as they had the dashboard annotation read permission.

Reply:
Ah, I thought it would just execute the `a.dashboard_id = 0` query and return only the organization dashboards. No worries then, let's leave the filter in there.

Reply:
This is roughly how this works:

``` mermaid
flowchart
    A["getAccessControlFilter()"] --> B{scope types}
    B --> |Organization| C["dashboard_id=0 (annotations without dashboard)"]
    B --> |Dashboard| D[filter to only visible dashboards]
    D --> E("join with 'OR'")
    C --> E
```